subsys/bluetooth/host/gatt.c

```diff
@@ -1211,73 +1225,110 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
 	cfg->value = 0U;
 }
 
-#if defined(CONFIG_BT_SETTINGS_CCC_STORE_ON_WRITE)
-static struct gatt_ccc_store {
-	struct bt_conn *conn_list[CONFIG_BT_MAX_CONN];
+#if defined(CONFIG_BT_SETTINGS_DELAYED_STORE)
+static struct gatt_delayed_store {
+	struct conn_list {
+		struct bt_conn *conn;
+
+		ATOMIC_DEFINE(flags, DELAYED_STORE_NUM_FLAGS);
+	} conn_list[CONFIG_BT_MAX_CONN];
 	struct k_work_delayable work;
-} gatt_ccc_store;
+} gatt_delayed_store;
 
-static bool gatt_ccc_conn_is_queued(struct bt_conn *conn)
+static bool gatt_delayed_store_is_queued(struct bt_conn *conn, enum delayed_store_flags flag)
 {
-	return (conn == gatt_ccc_store.conn_list[bt_conn_index(conn)]);
+	uint8_t index = bt_conn_index(conn);
+	struct conn_list *el = &gatt_delayed_store.conn_list[index];
+
+	return (conn == el->conn) && atomic_test_bit(el->flags, flag);
 }
 
-static void gatt_ccc_conn_unqueue(struct bt_conn *conn)
+static void gatt_delayed_store_dequeue(struct bt_conn *conn, enum delayed_store_flags flag)
 {
 	uint8_t index = bt_conn_index(conn);
+	struct conn_list *el = &gatt_delayed_store.conn_list[index];
 
-	if (gatt_ccc_store.conn_list[index] != NULL) {
-		bt_conn_unref(gatt_ccc_store.conn_list[index]);
-		gatt_ccc_store.conn_list[index] = NULL;
+	if (el->conn != NULL) {
+		__ASSERT_NO_MSG(el->conn == conn);
+		atomic_clear_bit(el->flags, flag);
+
+		if (atomic_get(el->flags) == 0) {
+			bt_conn_unref(el->conn);
+			el->conn = NULL;
+		}
 	}
 }
 
-static void gatt_ccc_conn_enqueue(struct bt_conn *conn)
+static void gatt_delayed_store_enqueue(struct bt_conn *conn, enum delayed_store_flags flag)
 {
-	if ((!gatt_ccc_conn_is_queued(conn)) &&
-	    bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
+	if (!conn) {
+		LOG_WRN("Not connected: cannot store client features to NVS");
+		return;
+	}
+
+	bool queued = gatt_delayed_store_is_queued(conn, flag);
+	bool bonded = bt_addr_le_is_bonded(conn->id, &conn->le.dst);
+
+	if (!queued && bonded) {
 		/* Store the connection with the same index it has in
 		 * the conns array
 		 */
-		gatt_ccc_store.conn_list[bt_conn_index(conn)] =
-			bt_conn_ref(conn);
+		uint8_t index = bt_conn_index(conn);
+		struct conn_list *el = &gatt_delayed_store.conn_list[index];
 
-		k_work_reschedule(&gatt_ccc_store.work, CCC_STORE_DELAY);
+		if (el->conn == NULL) {
+			el->conn = bt_conn_ref(conn);
+			atomic_set(el->flags, 0);
```

Comment:
There is hidden a type error here. `atomic_set` takes a pointer to a single `atomic_t`. `el->flags` is an array of `atomic_t`. This happens to work fine so long as `enum delayed_store_flags` has fewer than `ATOMIC_BITS` members, which admittedly is quite large; so maybe this case is fine by our code standard since we don't anticipate the flag count to grow that much?

Reply:
yeah I don't think we're gonna cross `ATOMIC_BITS`, is that 32 or something?